Complete mess with idleTimeout

**Jetty version(s)**
12.0.31

**Jetty Environment**
core, ee10

**HTTP version**
HTTP 1.1

**Java version/vendor** `(use: java -version)`
openjdk version "21.0.6" 2025-01-21 LTS
OpenJDK Runtime Environment Corretto-21.0.6.7.1 (build 21.0.6+7-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.6.7.1 (build 21.0.6+7-LTS, mixed mode, sharing)

**OS type/version**
Any Linux OS

**Description**
Maybe it would be useful for someone. This is how I figured out `Idle timeout expired: 30000/30000 ms` issue for websocket.

I have a legacy app that I'm bound to maintain. The app uses HTTP and WebSocket.
Some customers complain they experience accidental logouts every 30 seconds, as on websocket close app logs out the user. In logs I see these lines

```
Caused by: org.eclipse.jetty.websocket.core.exception.WebSocketTimeoutException: Connection Idle Timeout
Caused by: java.util.concurrent.TimeoutException: Idle timeout expired: 30000/30000 ms
```
In `JettyWebSocketServlet` derived class `configure` method I had set some value distinct from 30 seconds.
```java
aFactory.setIdleTimeout(Duration.ofMillis(timeout));
```
Regardless of this the issue happens every 30 seconds, but only for some customers.

In other app part there's http server created and the idle timeout set like this. The value also differs from 30 seconds.
```java
final HttpConnectionFactory httpConnectionFactory = new HttpConnectionFactory(httpConfig);
httpConnectionFactory.getHttpConfiguration().setIdleTimeout(timeoutMs);
```

Eventually what I figured out. Endpoints (jetty objects) actually hold idle timeout value, but it turns out it always setup multiple times regardless whether endpoint relates to websocket or httpconnection, doesn't matter.

This is logs
```
@4334e59f{s=END}]=>HttpChannelState@7920438c{handling=null, handled=true, send=LAST_COMPLETE, completed=true, request=GET@1d84cfe3 http://172.17.0.2:62616/oms/websocket HTTP/1.1}]
2026-01-16 17:36:45.745 [qtp1944295884-112] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 30000 on SocketChannelEndPoint@639a49e4[{l=/172.17.0.2:62616,r=/172.17.0.1:35200,OPEN,fill=-,flush=-,to=0/30000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 17:36:45.748 [qtp1944295884-108] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 30000 -> 120000 on SocketChannelEndPoint@639a49e4[{l=/172.17.0.2:62616,r=/172.17.0.1:35200,OPEN,fill=-,flush=-,to=1/120000}{io=0/0,kio=0,kro=1}]->[HttpConnection@65af77d5[p=HttpParser{s=HEADER,0 of -1},g=HttpGenerator@3a9978d1{s=START}]=>HttpChannelState@bfdd30f{handling=null, handled=false, send=SENDING, completed=false, request=GET@2b807330 http://172.17.0.2:62616/oms/websocket HTTP/1.1}]
2026-01-16 17:36:45.750 [qtp1944295884-108] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 120000 -> 60000 on SocketChannelEndPoint@639a49e4[{l=/172.17.0.2:62616,r=/172.17.0.1:35200,OPEN,fill=-,flush=-,to=3/60000}{io=0/0,kio=0,kro=1}]->[HttpConnection@65af77d5[p=HttpParser{s=END,0 of -1},g=HttpGenerator@3a9978d1{s=START}]=>HttpChannelState@bfdd30f{handling=Thread[#108,qtp1944295884-108,5,main], handled=false, send=SENDING, completed=false, request=GET@2b807330 http://172.17.0.2:62616/oms/websocket HTTP/1.1}]
2026-01-16 17:36:45.752 [qtp1944295884-108] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 60000 -> 30000 on SocketChannelEndPoint@639a49e4[{l=/172.17.0.2:62616,r=/172.17.0.1:35200,OPEN,fill=-,flush=-,to=1/30000}{io=0/0,kio=0,kro=1}]->[HttpConnection@65af77d5[p=HttpParser{s=END,0 of -1},g=HttpGenerator@3a9978d1{s=END}]=>HttpChannelState@bfdd30f{handling=null, handled=true, send=LAST_COMPLETE, completed=true, request=GET@2b807330 http://172.17.0.2:62616/oms/websocket HTTP/1.1}]
```
Here you can see three values distinct from zero. 120000 is for http, 60000 is for websocket I mentioned them above and 30000 that turns out hardcoded in the `org.eclipse.jetty.server.AbstractConnector`.  Looks like this flow is not stable and can change, sometimes I saw the end value is 120000.


<img width="849" height="353" alt="Image" src="https://github.com/user-attachments/assets/15b11a86-d4b3-47da-8312-361a1040e78f" />

From logs it's seen the same endpoint object `SocketChannelEndPoint@639a49e4` is assigned different timeouts ending up in 30000. That's the cause of the issue. So to resolve this I was bound to set the same value for the entities at first glance looking untied, like distinct connection instances providing different protocols.

<img width="857" height="299" alt="Image" src="https://github.com/user-attachments/assets/8c70b950-039c-45bc-8b64-fb16d452aea5" />
<img width="887" height="317" alt="Image" src="https://github.com/user-attachments/assets/73753569-5afa-4575-a033-c7ecc5dc226e" />

So be aware that http idle timeout will be exposed into your websocket connection instance. When I added above changes (marked  with 2, 3) I now have all the setup points set same value ending up with my value and not 30 seconds hardcoded... In this case only last two lines are for the same object as maybe due to same value check
```
2026-01-16 18:16:58.563 [qtp1944295884-138] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 120000 on SocketChannelEndPoint@d350052[{l=/172.17.0.2:62616,r=/172.17.0.1:33888,OPEN,fill=-,flush=-,to=3/120000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 18:16:58.712 [qtp1944295884-139] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 120000 on SocketChannelEndPoint@14e386b5[{l=/172.17.0.2:62616,r=/172.17.0.1:33890,OPEN,fill=-,flush=-,to=0/120000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 18:16:58.713 [qtp1944295884-140] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 120000 on SocketChannelEndPoint@42b28201[{l=/172.17.0.2:62616,r=/172.17.0.1:33896,OPEN,fill=-,flush=-,to=0/120000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 18:16:59.318 [qtp1944295884-141] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 120000 on SocketChannelEndPoint@1c0554c2[{l=/172.17.0.2:62616,r=/172.17.0.1:33910,OPEN,fill=-,flush=-,to=0/120000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 18:16:59.531 [qtp1944295884-111] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 0 -> 120000 on SocketChannelEndPoint@22bb1a2[{l=/172.17.0.2:62616,r=/172.17.0.1:33922,OPEN,fill=-,flush=-,to=0/120000}{io=0/0,kio=0,kro=0}]->[<null>]
2026-01-16 18:16:59.594 [qtp1944295884-113] DEBUG org.eclipse.jetty.io.IdleTimeout - Setting idle timeout 120000 -> 120000 on SocketChannelEndPoint@22bb1a2[{l=/172.17.0.2:62616,r=/172.17.0.1:33922,OPEN,fill=-,flush=-,to=60/120000}{io=0/0,kio=0,kro=1}]->[HttpConnection@12ae533a[p=HttpParser{s=END,0 of -1},g=HttpGenerator@fb3aa63{s=START}]=>HttpChannelState@50caac96{handling=Thread[#113,qtp1944295884-113,5,main], handled=false, send=SENDING, completed=false, request=GET@392361a6 http://172.17.0.2:62616/oms/websocket HTTP/1.1}]
```
I even saw they set idle timeout to new value in request handlers as well.  Now it works as expected.

--- a/jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketCoreSession.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketCoreSession.java
@@ -131,15 +131,20 @@ public class WebSocketCoreSession implements CoreSession, Dumpable
     @Override
     public Duration getIdleTimeout()
     {
-        return idleTimeout != null ? idleTimeout : WebSocketConstants.DEFAULT_IDLE_TIMEOUT;
+        if (idleTimeout != null)
+            return idleTimeout;
+        else if (connection != null)
+            return Duration.ofMillis(connection.getEndPoint().getIdleTimeout());
+        else
+            return WebSocketConstants.DEFAULT_IDLE_TIMEOUT;
     }
 
     @Override
     public void setIdleTimeout(Duration timeout)
     {
         idleTimeout = Objects.requireNonNull(timeout);
-        if (connection != null)
-            connection.getEndPoint().setIdleTimeout(timeout.toMillis());
+        if (!sessionState.isConnecting())
+            connection.getEndPoint().setIdleTimeout(idleTimeout.toMillis());
     }
 
     @Override
@@ -190,13 +195,6 @@ public class WebSocketCoreSession implements CoreSession, Dumpable
      */
     public void setWebSocketConnection(WebSocketConnection connection)
     {
-        // If the idle timeout is not initialized by a WebSocket
-        // configurator, then inherit that of the EndPoint; otherwise
-        // force the EndPoint to use the WebSocket configured idle timeout.
-        if (idleTimeout == null)
-            idleTimeout = Duration.ofMillis(connection.getEndPoint().getIdleTimeout());
-        else
-            connection.getEndPoint().setIdleTimeout(idleTimeout.toMillis());
         connection.setWriteTimeout(frameWriteTimeout.toMillis());
         extensionStack.setLastDemand(connection::demand);
         this.connection = connection;
@@ -383,6 +381,10 @@ public class WebSocketCoreSession implements CoreSession, Dumpable
     {
         if (LOG.isDebugEnabled())
             LOG.debug("onOpen() {}", this);
+
+        // Delay setting the idleTimeout until the session is opened.
+        if (idleTimeout != null)
+            connection.getEndPoint().setIdleTimeout(idleTimeout.toMillis());
 
         // Upgrade success
         sessionState.onConnected();

--- a/jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketSessionState.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketSessionState.java
@@ -42,6 +42,11 @@ public class WebSocketSessionState
     private byte _outgoingContinuation = OpCode.UNDEFINED;
     CloseStatus _closeStatus = null;
 
+    public boolean isConnecting()
+    {
+        return getState() == State.CONNECTING;
+    }
+
     public void onConnected()
     {
         try (AutoLock l = lock.lock())

--- a/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/WebSocketIdleTimeoutTest.java
+++ b/jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/WebSocketIdleTimeoutTest.java
@@ -1,0 +1,208 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.tests;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.jetty.server.HttpConfiguration;
+import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.websocket.api.Callback;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.exceptions.WebSocketTimeoutException;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.eclipse.jetty.websocket.common.WebSocketSession;
+import org.eclipse.jetty.websocket.core.WebSocketCoreSession;
+import org.eclipse.jetty.websocket.server.ServerWebSocketContainer;
+import org.eclipse.jetty.websocket.server.WebSocketUpgradeHandler;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import static org.eclipse.jetty.websocket.api.StatusCode.NORMAL;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class WebSocketIdleTimeoutTest
+{
+    private static final long IDLE_TIMEOUT_CONNECTOR = 10000;
+    private static final long IDLE_TIMEOUT_HTTP_CONFIG = 10001;
+    private static final long IDLE_TIMEOUT_CONTAINER = 10002;
+    private static final long IDLE_TIMEOUT_SESSION = 10003;
+
+    private final TimeoutEndpoint _serverEndpoint = new TimeoutEndpoint();
+    private Server _server;
+    private ServerConnector _connector;
+    private WebSocketClient _client;
+    private WebSocketUpgradeHandler _upgradeHandler;
+    private HttpConfiguration _httpConfig;
+
+    public void before(Runnable runnable) throws Exception
+    {
+        _server = new Server();
+        _httpConfig = new HttpConfiguration();
+        HttpConnectionFactory httpConnectionFactory = new HttpConnectionFactory(_httpConfig);
+        _connector = new ServerConnector(_server, httpConnectionFactory);
+        _server.addConnector(_connector);
+        _upgradeHandler = WebSocketUpgradeHandler.from(_server);
+        _server.setHandler(_upgradeHandler);
+        runnable.run();
+        _server.start();
+
+        _client = new WebSocketClient();
+        _client.start();
+    }
+
+    @AfterEach
+    public void after() throws Exception
+    {
+        _client.stop();
+        _server.stop();
+    }
+
+    public static class TimeoutEndpoint extends CloseTrackingEndpoint
+    {
+        private long idleTimeout = -1;
+
+        public void setIdleTimeout(long idleTimeout)
+        {
+            this.idleTimeout = idleTimeout;
+        }
+
+        @Override
+        public void onWebSocketOpen(Session session)
+        {
+            // Add an idle timeout listener to the session, which decides whether to close connection based on `_allowTimeout`.
+            if (idleTimeout >= 0)
+                session.setIdleTimeout(Duration.ofMillis(idleTimeout));
+            super.onWebSocketOpen(session);
+        }
+
+        @Override
+        public void onWebSocketText(String message)
+        {
+            if ("getIdleTimeout".equals(message))
+            {
+                // Use the endpoint directly to get the true idleTimeout value.
+                WebSocketSession webSocketSession = (WebSocketSession)getSession();
+                WebSocketCoreSession coreSession = (WebSocketCoreSession)webSocketSession.getCoreSession();
+                long idleTimeout = coreSession.getConnection().getEndPoint().getIdleTimeout();
+                getSession().sendText(Long.toString(idleTimeout), Callback.NOOP);
+            }
+        }
+    }
+
+    @Test
+    public void testWebSocketIdleTimeout() throws Exception
+    {
+        before(() ->
+        {
+            ServerWebSocketContainer container = _upgradeHandler.getServerWebSocketContainer();
+            container.addMapping("/", (req, resp, cb) -> _serverEndpoint);
+            container.setIdleTimeout(Duration.ofMillis(1000));
+        });
+
+        EventSocket clientEndpoint = new EventSocket();
+        _client.connect(clientEndpoint, URI.create("ws://localhost:" + _connector.getLocalPort()));
+        assertTrue(_serverEndpoint.openLatch.await(5, TimeUnit.SECONDS));
+
+        // Allow the timeout listener to close the connection.
+        assertTrue(_serverEndpoint.closeLatch.await(5, TimeUnit.SECONDS));
+        assertThat(_serverEndpoint.error.get(), instanceOf(WebSocketTimeoutException.class));
+    }
+
+    @Test
+    public void testConnectorIdleTimeout() throws Exception
+    {
+        before(() ->
+        {
+            ServerWebSocketContainer container = _upgradeHandler.getServerWebSocketContainer();
+            container.addMapping("/", (req, resp, cb) -> _serverEndpoint);
+
+            _connector.setIdleTimeout(IDLE_TIMEOUT_CONNECTOR);
+        });
+
+        assertIdleTimeoutAndClose(IDLE_TIMEOUT_CONNECTOR);
+    }
+
+    @Test
+    public void testHttpConfigurationIdleTimeout() throws Exception
+    {
+        before(() ->
+        {
+            ServerWebSocketContainer container = _upgradeHandler.getServerWebSocketContainer();
+            container.addMapping("/", (req, resp, cb) -> _serverEndpoint);
+
+            _connector.setIdleTimeout(IDLE_TIMEOUT_CONNECTOR);
+            _httpConfig.setIdleTimeout(IDLE_TIMEOUT_HTTP_CONFIG);
+        });
+
+        assertIdleTimeoutAndClose(IDLE_TIMEOUT_CONNECTOR);
+    }
+
+    @Test
+    public void testWebSocketContainerIdleTimeout() throws Exception
+    {
+        before(() ->
+        {
+            ServerWebSocketContainer container = _upgradeHandler.getServerWebSocketContainer();
+            container.addMapping("/", (req, resp, cb) -> _serverEndpoint);
+
+            _connector.setIdleTimeout(IDLE_TIMEOUT_CONNECTOR);
+            _httpConfig.setIdleTimeout(IDLE_TIMEOUT_HTTP_CONFIG);
+            container.setIdleTimeout(Duration.ofMillis(IDLE_TIMEOUT_CONTAINER));
+        });
+
+        assertIdleTimeoutAndClose(IDLE_TIMEOUT_CONTAINER);
+    }
+
+    @Test
+    public void testWebSocketSessionIdleTimeout() throws Exception
+    {
+        before(() ->
+        {
+            ServerWebSocketContainer container = _upgradeHandler.getServerWebSocketContainer();
+            _connector.setIdleTimeout(IDLE_TIMEOUT_CONNECTOR);
+            _httpConfig.setIdleTimeout(IDLE_TIMEOUT_HTTP_CONFIG);
+            container.setIdleTimeout(Duration.ofMillis(IDLE_TIMEOUT_CONTAINER));
+            _serverEndpoint.setIdleTimeout(IDLE_TIMEOUT_SESSION);
+
+            container.addMapping("/", (req, resp, cb) -> _serverEndpoint);
+        });
+
+        assertIdleTimeoutAndClose(IDLE_TIMEOUT_SESSION);
+    }
+
+    private void assertIdleTimeoutAndClose(long expectedIdleTimeout) throws Exception
+    {
+        EventSocket clientEndpoint = new EventSocket();
+        _client.connect(clientEndpoint, URI.create("ws://localhost:" + _connector.getLocalPort()));
+        assertTrue(_serverEndpoint.openLatch.await(5, TimeUnit.SECONDS));
+        assertTrue(clientEndpoint.openLatch.await(5, TimeUnit.SECONDS));
+
+        clientEndpoint.session.sendText("getIdleTimeout", Callback.NOOP);
+        String idleTimeout = clientEndpoint.textMessages.poll(5, TimeUnit.SECONDS);
+        assertThat(idleTimeout, equalTo(Long.toString(expectedIdleTimeout)));
+
+        _serverEndpoint.getSession().close();
+        assertTrue(_serverEndpoint.closeLatch.await(5, TimeUnit.SECONDS));
+        assertThat(_serverEndpoint.closeCode, equalTo(NORMAL));
+    }
+}

